mavros/src/plugins/hil.cpp

```diff
@@ -0,0 +1,320 @@
+/**
+ * @brief HilControls plugin
+ * @file hil_controls.cpp
+ * @author Pavel Vechersky <pvechersky@student.ethz.ch>
+ * @author Beat Küng <beat-kueng@gmx.net>
+ * @author Nuno Marques <n.marques21@hotmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2016,2017 Pavel Vechersky, Beat Küng, Nuno Marques.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+
+#include <mavros_msgs/HilControls.h>
+#include <mavros_msgs/HilActuatorControls.h>
+#include <mavros_msgs/HilStateQuaternion.h>
+#include <mavros_msgs/HilGPS.h>
+#include <mavros_msgs/HilSensor.h>
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Hil plugin
+ */
+class HilPlugin : public plugin::PluginBase {
+public:
+	HilPlugin() : PluginBase(),
+		hil_nh("~hil")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+		last_time = ros::Time(0.0);
+		period = ros::Duration(0.025);	// 40hz
+
+		hilStateQuaternion_sub = hil_nh.subscribe("hil_state", 10, &HilPlugin::state_quat_cb, this);
+		hilGPS_sub = hil_nh.subscribe("hilgps", 10, &HilPlugin::gps_cb, this);
+		hilSensor_sub = hil_nh.subscribe("imu_ned", 10, &HilPlugin::sensor_cb, this);
+
+		hil_controls_pub = hil_nh.advertise<mavros_msgs::HilControls>("hil_controls", 10);
+		hil_actuator_controls_pub = hil_nh.advertise<mavros_msgs::HilActuatorControls>("hil_actuator_controls", 10);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			       make_handler(&HilPlugin::handle_hil_controls),
+			       make_handler(&HilPlugin::handle_hil_actuator_controls),
+		};
+	}
+
+private:
+	ros::NodeHandle hil_nh;
+
+	ros::Publisher hil_controls_pub;
+	ros::Publisher hil_actuator_controls_pub;
+
+	ros::Subscriber hilStateQuaternion_sub;
+	ros::Subscriber hilGPS_sub;
+	ros::Subscriber hilSensor_sub;
+
+	ros::Time last_time;
+	ros::Duration period;
+
+	/* -*- rx handlers -*- */
+
+	void handle_hil_controls(const mavlink::mavlink_message_t *msg, mavlink::common::msg::HIL_CONTROLS &hil_controls) {
+		// Throttle incoming messages to 40hz
+		if ((ros::Time::now() - last_time) < period) {
```

Comment:
Well that is @mzahana implementation. @mzahana can you explain why you throttle the message here? Any reason to do it?